File: software/web/src/api/services/sensor_data.ts
import { SensorData } from '../../types/sensor_data';
import { axios } from '../axios';
import { API_SENSOR_DATA_HISTORICAL } from '../urls';


interface ResponseSensorDataHistorical {
    data: {
        sensor_data: Record<string, string>[];
    };
    status: string,
}

export const getHistorical = async (startDate: string, endDate: string): Promise<SensorData[]> => {
    const queryParameters = `?start_date=${startDate}&end_date=${endDate}`;

    return axios.get<any, ResponseSensorDataHistorical>(`${API_SENSOR_DATA_HISTORICAL}${queryParameters}`).then((data) => {
        if (data) {
            return data.data.sensor_data.map(i => {
                return {
                    "controller": i.controller,
                    "controllerName": i.controller_name,
                    "area": i.area,
                    "areaName": i.area_name,
                    "humidity": i.humidity,
                    "temperature": i.temperature,
                    "raining": i.raining,
                    "date": i.date,
                } as unknown as SensorData
            });
        }

        return Promise.reject();
    })
}

File: software/server/irrigation/api/dto/sensor_data.py
""" DTO - sensor data historical schemas for rest api
"""
from marshmallow import Schema, pre_load, fields


class ApiHistoricalSensorDataSchema(Schema):
    controller_id = fields.Integer(
        allow_none=True,
        data_key='controller',
        error_messages={
            'invalid': '`controller` must be an Integer.'
        })
    area_id = fields.Integer(
        allow_none=True,
        data_key='area',
        error_messages={
            'invalid': '`area` must be a Integer.'
        })
    start_date = fields.String(
        allow_none=True,
        data_key='start_date',
        error_messages={
            'invalid': '`start_date` must be a String.'
        })
    end_date = fields.String(
        allow_none=True,
        data_key='end_date',
        error_messages={
            'invalid': '`start_date` must be a String.'
        })
    humidity = fields.Float(
        allow_none=True,
        data_key='humidity',
        error_messages={
            'invalid': '`humidity` must be a Float.'
        })
    temperature = fields.Float(
        allow_none=True,
        data_key='temperature',
        error_messages={
            'invalid': '`temperature` must be a Float.'
        })
    raining = fields.Boolean(
        allow_none=True,
        data_key='raining',
        error_messages={
            'invalid': '`raining` must be a Boolean.'
        })


class ApiSensorDataSchema(Schema):
    controller_id = fields.Integer(
        allow_none=True,
        data_key='controller',
        error_messages={
            'invalid': '`controller` must be an Integer.'
        })
    controller = fields.String(
        allow_none=True,
        data_key='controller_name',
        error_messages={
            'invalid': '`controller_name` must be a String.'
        })
    area_id = fields.Integer(
        allow_none=True,
        data_key='area',
        error_messages={
            'invalid': '`area` must be a Integer.'
        })
    area = fields.String(
        allow_none=True,
        data_key='area_name',
        error_messages={
            'invalid': '`area_name` must be a String.'
        })
    humidity = fields.Float(
        allow_none=True,
        data_key='humidity',
        error_messages={
            'invalid': '`humidity` must be a Float.'
        })
    temperature = fields.Float(
        allow_none=True,
        data_key='temperature',
        error_messages={
            'invalid': '`temperature` must be a Float.'
        })
    raining = fields.Boolean(
        allow_none=True,
        data_key='raining',
        error_messages={
            'invalid': '`raining` must be a Boolean.'
        })
    date = fields.String(
        allow_none=True,
        data_key='date',
        error_messages={
            'invalid': '`date` must be a String.'
        })
